File: scripts/generate_gui.lua
-- scripts/generate_gui.lua
-- Converts gui.rbxmx into a Lua module that builds the interface with
-- [`Instance.new`](https://create.roblox.com/docs/reference/engine/functions/Instance/new)
-- calls. This avoids loading the XML file directly at runtime.
-- The output is written to src/GeneratedGui.lua.

local function readFile(path)
    local f = assert(io.open(path, "r"))
    local data = f:read("*a")
    f:close()
    return data
end

local function parseAttributes(str)
    local attrs = {}
    for k, v in string.gmatch(str, "([%w_:%-]+)%s*=%s*\"(.-)\"") do
        attrs[k] = v
    end
    return attrs
end

local function parse(xml)
    local pos = 1
    local root = {children = {}}
    local stack = {root}
    while true do
        local s, e, closing, name, attrstr, empty = xml:find("<(/?)([%w_:%-]+)(.-)(/?)>", pos)
        if not s then break end
        local text = xml:sub(pos, s - 1)
        if text:match("%S") then
            local top = stack[#stack]
            top.children[#top.children + 1] = {text = text}
        end
        if closing == "/" then
            table.remove(stack)
        else
            local node = {name = name, attrs = parseAttributes(attrstr), children = {}}
            local parent = stack[#stack]
            parent.children[#parent.children + 1] = node
            if empty ~= "/" then
                stack[#stack + 1] = node
            end
        end
        pos = e + 1
    end
    return root.children[1]
end

local function propValue(node)
    local t = node.name
    local val = node.children[1] and node.children[1].text or ""
    if t == "string" or t == "BinaryString" then
        return string.format("%q", val)
    elseif t == "Content" then
        local urlNode = node.children[1]
        if urlNode and urlNode.name == "url" then
            val = urlNode.text or ""
        end
        return string.format("%q", val)
    elseif t == "bool" then
        return tostring(val == "true")
    elseif t == "int" or t == "int64" or t == "float" or t == "double" or t == "token" then
        return tonumber(val) or 0
    elseif t == "Vector2" then
        local x, y = 0, 0
        for _, c in ipairs(node.children) do
            if c.name == "X" then x = tonumber(c.children[1].text) end
            if c.name == "Y" then y = tonumber(c.children[1].text) end
        end
        return string.format("Vector2.new(%s,%s)", x, y)
    elseif t == "UDim2" then
        local xs, xo, ys, yo = 0, 0, 0, 0
        for _, c in ipairs(node.children) do
            if c.name == "XS" or c.name == "XScale" then xs = tonumber(c.children[1].text) end
            if c.name == "XO" or c.name == "XOffset" then xo = tonumber(c.children[1].text) end
            if c.name == "YS" or c.name == "YScale" then ys = tonumber(c.children[1].text) end
            if c.name == "YO" or c.name == "YOffset" then yo = tonumber(c.children[1].text) end
        end
        local baseW, baseH = 1920, 1080
        xs = xs + xo / baseW
        ys = ys + yo / baseH
        return string.format("UDim2.fromScale(%s,%s)", xs, ys)
    elseif t == "Color3" then
        local r, g, b = 0, 0, 0
        for _, c in ipairs(node.children) do
            if c.name == "R" then r = tonumber(c.children[1].text) end
            if c.name == "G" then g = tonumber(c.children[1].text) end
            if c.name == "B" then b = tonumber(c.children[1].text) end
        end
        return string.format("Color3.new(%s,%s,%s)", r, g, b)
    elseif t == "UDim" then
        local s, o = 0, 0
        for _, c in ipairs(node.children) do
            if c.name == "S" or c.name == "Scale" then s = tonumber(c.children[1].text) end
            if c.name == "O" or c.name == "Offset" then o = tonumber(c.children[1].text) end
        end
        local base = 1000
        s = s + o / base
        return string.format("UDim.new(%s,0)", s)
    elseif t == "Font" then
        local family = ""
        local weight = 400
        local style = "Normal"
        for _, c in ipairs(node.children) do
            if c.name == "Family" then
                local urlNode = c.children[1]
                family = urlNode and urlNode.text or ""
            elseif c.name == "Weight" then
                weight = tonumber(c.children[1] and c.children[1].text) or 400
            elseif c.name == "Style" then
                style = c.children[1] and c.children[1].text or "Normal"
            end
        end
        local weightEnum = "Enum.FontWeight.Regular"
        if weight and weight >= 700 then
            weightEnum = "Enum.FontWeight.Bold"
        end
        local styleEnum = "Enum.FontStyle." .. style
        return string.format("Font.new(%q, %s, %s)", family, weightEnum, styleEnum)
    end
    return string.format("%q", val)
end

local outLines = {}
local index = 0
local function genItem(node, parent)
    if node.name ~= "Item" then return end
    index = index + 1
    local var = "obj" .. index
    outLines[#outLines + 1] = string.format("    local %s = Instance.new(%q)", var, node.attrs.class)
    for _, child in ipairs(node.children) do
        if child.name == "Properties" then
            for _, p in ipairs(child.children) do
                if p.name then
                    local propName = p.attrs.name
                    if propName ~= "SourceAssetId"
                        and propName ~= "Tags"
                        and propName ~= "LocalizationMatchIdentifier"
                        and propName ~= "LocalizationMatchedSourceText" then
                        outLines[#outLines + 1] =
                            string.format("    %s.%s = %s", var, propName, propValue(p))
                    end
                end
            end
        elseif child.name == "Item" then
            genItem(child, var)
        end
    end
    if parent then
        outLines[#outLines + 1] = string.format("    %s.Parent = %s", var, parent)
    end
    if node.attrs.class == "TextLabel" then
        local cornerVar = "corner" .. index
        outLines[#outLines + 1] = string.format("    local %s = Instance.new(\"UICorner\")", cornerVar)
        outLines[#outLines + 1] = string.format("    %s.CornerRadius = UDim.new(0.05,0)", cornerVar)
        outLines[#outLines + 1] = string.format("    %s.Parent = %s", cornerVar, var)
    end
end

local xml = readFile("gui.rbxmx")
local tree = parse(xml)

outLines[#outLines + 1] = "local function createGui(parent)"
outLines[#outLines + 1] = "    local screenGui = Instance.new(\"ScreenGui\")"
outLines[#outLines + 1] = "    screenGui.Name = \"InventoryUI\""
outLines[#outLines + 1] = "    screenGui.IgnoreGuiInset = true"
outLines[#outLines + 1] = "    screenGui.ResetOnSpawn = false"
outLines[#outLines + 1] = "    if parent then screenGui.Parent = parent end"

for _, child in ipairs(tree.children) do
    if child.name == "Item" then
        genItem(child, "screenGui")
    end
end

outLines[#outLines + 1] = "    return screenGui"
outLines[#outLines + 1] = "end"
outLines[#outLines + 1] = "return createGui"

local outFile = assert(io.open("src/GeneratedGui.lua", "w"))
outFile:write(table.concat(outLines, "\n"))
outFile:close()

print("Generated src/GeneratedGui.lua")

File: scripts/generate_gui.py
import xml.etree.ElementTree as ET

SKIP_PROPERTIES = {
    "SourceAssetId",
    "Tags",
    "LocalizationMatchIdentifier",
    "LocalizationMatchedSourceText",
    "AttributesSerialize",
    "Capabilities",
    "DefinesCapabilities",
    "RootLocalizationTable",
    "HoverHapticEffect",
    "HoverImage",
    "PressHapticEffect",
    "PressedImage",
    "ResetOnSpawn",
}

BASE_W = 1920
BASE_H = 1080

BASE_UDIM = 1000

index = 0
lines = []


def prop_value(node):
    tag = node.tag
    val = (node.text or "").strip()
    if tag in {"string", "BinaryString", "Content"}:
        return repr(val)
    if tag == "bool":
        return "true" if val == "true" else "false"
    if tag in {"int", "int64", "float", "double", "token"}:
        return val
    if tag == "Ref":
        return "nil"
    if tag == "Vector2":
        x = y = "0"
        for c in node:
            if c.tag == "X":
                x = c.text or "0"
            elif c.tag == "Y":
                y = c.text or "0"
        return f"Vector2.new({x},{y})"
    if tag == "UDim2":
        xs = xo = ys = yo = 0.0
        for c in node:
            if c.tag in ("XS", "XScale"):
                xs = float(c.text or 0)
            elif c.tag in ("XO", "XOffset"):
                xo = float(c.text or 0)
            elif c.tag in ("YS", "YScale"):
                ys = float(c.text or 0)
            elif c.tag in ("YO", "YOffset"):
                yo = float(c.text or 0)
        xs = xs + xo / BASE_W
        ys = ys + yo / BASE_H
        return f"UDim2.fromScale({xs},{ys})"
    if tag == "Color3":
        r = g = b = 0.0
        for c in node:
            if c.tag == "R":
                r = float(c.text or 0)
            elif c.tag == "G":
                g = float(c.text or 0)
            elif c.tag == "B":
                b = float(c.text or 0)
        return f"Color3.new({r},{g},{b})"
    if tag == "UDim":
        s = o = 0.0
        for c in node:
            if c.tag in ("S", "Scale"):
                s = float(c.text or 0)
            elif c.tag in ("O", "Offset"):
                o = float(c.text or 0)
        s = s + o / BASE_UDIM
        return f"UDim.new({s},0)"
    if tag == "Font":
        family = ""
        weight = 400
        style = "Normal"
        for c in node:
            if c.tag == "Family":
                family = ""
                for sub in c:
                    if sub.tag.lower() == "url":
                        family = (sub.text or "").strip()
                        break
            elif c.tag == "Weight":
                try:
                    weight = int(c.text or 400)
                except ValueError:
                    weight = 400
            elif c.tag == "Style":
                style = (c.text or "Normal").strip()
        weight_enum = "Enum.FontWeight.Regular"
        if weight >= 700:
            weight_enum = "Enum.FontWeight.Bold"
        style_enum = f"Enum.FontStyle.{style}"
        return f"Font.new({repr(family)}, {weight_enum}, {style_enum})"
    if tag == "Rect2D":
        min_x = min_y = max_x = max_y = 0
        for c in node:
            if c.tag == "min":
                for m in c:
                    if m.tag == "X":
                        min_x = int(m.text or 0)
                    elif m.tag == "Y":
                        min_y = int(m.text or 0)
            elif c.tag == "max":
                for m in c:
                    if m.tag == "X":
                        max_x = int(m.text or 0)
                    elif m.tag == "Y":
                        max_y = int(m.text or 0)
        return f"Rect.new({min_x},{min_y},{max_x},{max_y})"
    return repr(val)


def gen_item(item, parent):
    global index
    if item.tag != "Item":
        return
    index += 1
    var = f"obj{index}"
    lines.append(f"    local {var} = Instance.new(\"{item.attrib['class']}\")")
    for child in item:
        if child.tag == "Properties":
            for p in child:
                prop_name = p.attrib.get('name')
                if not prop_name or prop_name in SKIP_PROPERTIES:
                    continue
                lines.append(f"    {var}.{prop_name} = {prop_value(p)}")
        elif child.tag == "Item":
            gen_item(child, var)
    if parent:
        lines.append(f"    {var}.Parent = {parent}")
    if item.attrib.get('class') == "TextLabel":
        corner = f"corner{index}"
        lines.append(f"    local {corner} = Instance.new(\"UICorner\")")
        lines.append(f"    {corner}.CornerRadius = UDim.new(0.05,0)")
        lines.append(f"    {corner}.Parent = {var}")


def main():
    tree = ET.parse('gui.rbxmx')
    root = tree.getroot()
    first_item = None
    for child in root:
        if child.tag == 'Item':
            first_item = child
            break
    if first_item is None:
        raise RuntimeError('No Item found in gui.rbxmx')
    lines.append("local function createGui(parent)")
    lines.append("    local screenGui = Instance.new(\"ScreenGui\")")
    for child in first_item:
        if child.tag == "Properties":
            for p in child:
                prop_name = p.attrib.get('name')
                if not prop_name or prop_name in SKIP_PROPERTIES:
                    continue
                if prop_name == "Name":
                    lines.append("    screenGui.Name = \"InventoryUI\"")
                else:
                    lines.append(f"    screenGui.{prop_name} = {prop_value(p)}")
        elif child.tag == "Item":
            gen_item(child, "screenGui")
    lines.append("    screenGui.IgnoreGuiInset = true")
    reset_val = None
    for child in first_item:
        if child.tag == "Properties":
            for p in child:
                if p.attrib.get('name') == "ResetOnSpawn":
                    reset_val = (p.text or "").strip()
                    break
    if reset_val != "false":
        lines.append("    screenGui.ResetOnSpawn = false")
    lines.append("    if parent then screenGui.Parent = parent end")

    lines.append("    return screenGui")
    lines.append("end")
    lines.append("return createGui")

    with open('src/GeneratedGui.lua', 'w') as f:
        f.write("\n".join(lines))


if __name__ == '__main__':
    main()
